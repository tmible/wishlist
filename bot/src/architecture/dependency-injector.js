/** @module Модуль внедрения зависимостей */

/** @typedef {import('./injection-token.js').InjectionToken} InjectionToken */

/**
 * Отображение токенов внедрения в предоставляемые значения
 * @type {Map<InjectionToken, unknown>}
 * @see {@link import('./injection-token.js')} InjectionToken
 */
const providers = new Map();

/**
 * Ассоциация значения с токеном внедрения
 * @function provide
 * @param {InjectionToken} token Токен внедрения
 * @param {unknown} value Предоставляемое значение
 * @see {@link import('./injection-token.js')} InjectionToken
 */
export const provide = (token, value) => {
  providers.set(token, value);
};

/**
 * Предоставление значения по токену внедрения
 * @function inject
 * @param {InjectionToken} token Токен внедрения
 * @returns {unknown} Предоставляемое значение, ассоциированное с токеном
 * @see {@link import('./injection-token.js')} InjectionToken
 */
export const inject = (token) => providers.get(token);
