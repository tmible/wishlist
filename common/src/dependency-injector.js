/** @module Модуль внедрения зависимостей */

/** @typedef {string} InjectionToken */

/**
 * Отображение токенов внедрения в предоставляемые значения
 * @type {Map<InjectionToken, unknown>}
 */
const providers = new Map();

/**
 * Ассоциация значения с токеном внедрения
 * @function provide
 * @param {InjectionToken} token Токен внедрения
 * @param {unknown} value Предоставляемое значение
 */
export const provide = (token, value) => {
  providers.set(token, value);
};

/**
 * Предоставление значения по токену внедрения
 * @function inject
 * @param {InjectionToken} token Токен внедрения
 * @returns {unknown} Предоставляемое значение, ассоциированное с токеном
 */
export const inject = (token) => providers.get(token);
