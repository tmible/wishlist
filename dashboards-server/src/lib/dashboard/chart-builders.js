import { inject } from '@tmible/wishlist-common/dependency-injector';
import { browser } from '$app/environment';
import { PERIOD } from '$lib/constants/period.const.js';
import { ThemeService } from '$lib/theme-service-injection-token.js';

/** @typedef {import('chart.js').ChartConfiguration} ChartConfiguration */
/** @typedef {import('./domain.js').Dashboard} Dashboard */
/** @typedef {import('./domain.js').LineDashboard} LineDashboard */

/** @module Построение конфигов для графиков дашбордов */

/* eslint-disable jsdoc/check-template-names, jsdoc/valid-types --
  Валидный TS, но невалидный JSDoc */
/**
 * @template {Dashboard} [D=Dashboard]
 * @typedef {(D['charts'] extends Map<unknown, infer C> ? C : never)[]} DashboardChart
 */
/* eslint-enable jsdoc/check-template-names, jsdoc/valid-types */

/**
 * Цвет координатной сетки в светлой теме
 * @constant {string}
 */
const GRID_LIGHT_COLOR = 'rgba(0,0,0,0.1)';

/**
 * Получение hex цвета из CSS переменной
 * @function getColorFromStyles
 * @param {string} colorName Название цвета
 * @returns {string} hex цвет
 */
const getColorFromStyles = (colorName) => browser && getComputedStyle(
  document.documentElement,
).getPropertyValue(
  `--color-${colorName}`,
);

/**
 * Цвет успеха
 * @type {string}
 */
const successColor = getColorFromStyles('success');

/**
 * Цвет ошибки
 * @type {string}
 */
const errorColor = getColorFromStyles('error');

/**
 * Нахождение медианы в массиве точек графика
 * @function median
 * @param {object} ctx Контекст графика // no type declared for this parameter in chart.js docs
 * @returns {number} Медиана точек графика
 */
const median = (ctx) => {
  const values = ctx.chart.data.datasets[0]?.data.slice().sort((a, b) => a - b) ?? [];
  const half = Math.floor(values.length / 2);
  return (values.length % 2 === 1 ? values[half] : (values[half - 1] + values[half]) / 2);
};

/**
 * Построение конфига для дашборда с графиком в виде кольца
 * @function line
 * @param {string[]} labels Подписи к линиям на графике
 * @param {DashboardChart<LineDashboard>[]} charts Данные для построения линий
 * @param {number} periodStart Таймштамп начала выбранного для отображения периода
 * @param {Partial<ChartConfiguration['options']['scales']['y']>} yScaleOptions
 *   Дополнительные опции для настройки отображения оси Oy
 * @param {string} tooltipFormat Формат даты в тултипе
 * @param {PERIOD} periodSelected Выбранный для отображения период
 * @returns {ChartConfiguration} Конфигурация графика chart.js
 */
// eslint-disable-next-line max-lines-per-function -- большой конфиг, нет смысла разбивать
const line = (labels, charts, periodStart, yScaleOptions, tooltipFormat, periodSelected) => {
  const gridOptions = {
    grid: {
      color: inject(ThemeService).isDarkTheme() ?
        getColorFromStyles('secondary') :
        GRID_LIGHT_COLOR,
    },
  };

  return {
    type: 'line',
    data: {
      labels,
      datasets: charts.map(({ label, data }) => ({
        label,
        data: data?.map(({ value }) => value) ?? [],
        pointHitRadius: 10,
      })),
    },
    options: {
      scales: {
        x: {
          type: 'time',
          min: periodStart,
          max: Date.now(),
          ...gridOptions,
          ticks: {
            maxRotation: 0,
            autoSkipPadding: 20,
          },
          time: {
            displayFormats: {
              minute: 'HH:mm',
              hour: 'HH:00',
              day: 'DD.MM',
            },
            minUnit: periodSelected >= PERIOD.WEEK ?
              'day' :
              (periodSelected >= PERIOD.SIX_HOURS ?
                'hour' :
                'minute'),
            tooltipFormat,
          },
        },
        y: {
          min: 0,
          ...gridOptions,
          ...yScaleOptions,
        },
      },
      plugins: {
        annotation: {
          annotations: {
            annotation: {
              type: 'line',
              borderColor: getColorFromStyles('base-content'),
              borderDash: [ 6, 6 ],
              borderDashOffset: 0,
              borderWidth: 2,
              label: {
                display: true,
                content: (ctx) => `Медиана: ${median(ctx)}`,
                position: 'end',
              },
              scaleID: 'y',
              value: (ctx) => median(ctx),
            },
          },
        },
      },
    },
  };
};

/**
 * Построение конфига для дашборда с графиком в виде кольца
 * @function doughnut
 * @param {string[]} labels Подписи к секторам кольца
 * @param {number[]} data Данные для отображения
 * @param {string[]} backgroundColor Цвет фона для каждого сектора кольца
 * @param {string} borderColor Цвет границ секторов
 * @param {string} cutout Отношение внутреннего радиуса кольцо ко внешнему в процентах
 * @returns {ChartConfiguration} Конфигурация графика chart.js
 */
const doughnut = (labels, data, backgroundColor, borderColor, cutout) => ({
  type: 'doughnut',
  data: {
    labels,
    datasets: [{
      data,
      backgroundColor,
    }],
  },
  options: {
    borderColor,
    cutout,
    plugins: {
      legend: {
        display: false,
      },
      tooltip: {
        callbacks: {
          label: (context) => ` ${(context.raw * 100).toFixed(2)}%`,
        },
      },
    },
  },
});

/**
 * Отображение типов дашбордов в функции построения конфигов графиков
 * @constant {{
 *   [key: string]: (
 *     charts: DashboardChart[],
 *     periodStart: number,
 *     periodSelected: PERIOD,
 *  ) => ChartConfiguration
 * }}
 */
export const ChartBuilders = {
  time: (charts, periodStart, periodSelected) => line(
    charts[0]?.data?.map((item) => item.timestamp) ?? [],
    charts,
    periodStart,
    { title: { display: true, text: 'мс' } },
    'D.MM.YYYY H:mm:ss.SSS',
    periodSelected,
  ),
  activeUsers: (charts, periodStart, periodSelected) => line(
    charts[0]?.data?.map((item) => item.timestamp) ?? [],
    charts,
    periodStart,
    {},
    'D.MM.YYYY',
    periodSelected,
  ),
  successRate: (charts) => {
    const successRate = charts[0].data;
    return doughnut(
      [ 'success rate', 'error rate' ],
      successRate ? [ successRate, (1 - successRate) ] : [ 1 ],
      [ successColor, errorColor ],
      successRate ? getColorFromStyles('base-100') : successColor,
      successRate ? '50%' : '100%',
    );
  },
  authenticationFunnel: (charts) => {
    const authenticationFunnel = charts[0].data;
    return doughnut(
      [ 'аутентификации', 'уходы' ],
      authenticationFunnel ? [ authenticationFunnel, (1 - authenticationFunnel) ] : [ 1 ],
      [ successColor, errorColor ],
      authenticationFunnel ? getColorFromStyles('base-100') : successColor,
      authenticationFunnel ? '50%' : '100%',
    );
  },
};
