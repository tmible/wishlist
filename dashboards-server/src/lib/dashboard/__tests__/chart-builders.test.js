import { inject } from '@tmible/wishlist-common/dependency-injector';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { ChartBuilders } from '../chart-builders.js';

vi.mock('@tmible/wishlist-common/dependency-injector');
vi.mock('$app/environment', () => ({ browser: true }));
vi.hoisted(() => {
  vi.stubGlobal(
    'getComputedStyle',
    vi.fn().mockReturnValue({
      getPropertyValue: vi.fn((argument) => argument),
    }),
  );
  vi.stubGlobal('document', { documentElement: 'documentElement' });
});

describe('dashboard / chart builders', () => {
  afterEach(() => {
    vi.clearAllMocks();
  });

  for (const chart of [ 'time', 'activeUsers', 'rps' ]) {
    describe('for light theme', () => {
      beforeEach(() => {
        vi.mocked(inject).mockReturnValueOnce({ isDarkTheme: () => false });
        vi.spyOn(Date, 'now').mockReturnValueOnce('now');
      });

      it(`should build ${chart} chart with data`, () => {
        expect(
          ChartBuilders[chart](
            [
              { label: 'chart 1', data: [{ timestamp: 2, value: 3 }] },
              { label: 'chart 2', data: undefined },
            ],
            0,
            1,
          ),
        ).toMatchSnapshot();
      });

      it(`should build ${chart} chart without data`, () => {
        expect(ChartBuilders[chart]([], 0, 1)).toMatchSnapshot();
      });
    });

    describe('for dark theme', () => {
      beforeEach(() => {
        vi.mocked(inject).mockReturnValueOnce({ isDarkTheme: () => true });
        vi.spyOn(Date, 'now').mockReturnValueOnce('now');
      });

      it(`should build ${chart} chart with data`, () => {
        expect(
          ChartBuilders[chart](
            [
              { label: 'chart 1', data: [{ timestamp: 2, value: 3 }] },
              { label: 'chart 2', data: undefined },
            ],
            0,
            1,
          ),
        ).toMatchSnapshot();
      });

      it(`should build ${chart} chart without data`, () => {
        expect(ChartBuilders[chart]([], 0, 1)).toMatchSnapshot();
      });
    });
  }

  for (const chart of [ 'successRate', 'authenticationFunnel' ]) {
    it(`should build ${chart} chart with data`, () => {
      expect(ChartBuilders[chart]([{ data: 0.5 }])).toMatchSnapshot();
    });

    it(`should build ${chart} chart without data`, () => {
      expect(ChartBuilders[chart]([{ data: undefined }])).toMatchSnapshot();
    });
  }
});
