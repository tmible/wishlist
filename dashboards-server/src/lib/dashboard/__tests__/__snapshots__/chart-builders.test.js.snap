// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dashboard / chart builders > for dark theme > should build activeUsers chart with data 1`] = `
{
  "data": {
    "datasets": [
      {
        "data": [
          3,
        ],
        "label": "chart 1",
        "pointHitRadius": 10,
      },
      {
        "data": [],
        "label": "chart 2",
        "pointHitRadius": 10,
      },
    ],
    "labels": [
      2,
    ],
  },
  "options": {
    "plugins": {
      "annotation": {
        "annotations": {
          "annotation": {
            "borderColor": "hsl(--foreground)",
            "borderDash": [
              6,
              6,
            ],
            "borderDashOffset": 0,
            "borderWidth": 2,
            "label": {
              "content": [Function],
              "display": true,
              "position": "end",
            },
            "scaleID": "y",
            "type": "line",
            "value": [Function],
          },
        },
      },
    },
    "scales": {
      "x": {
        "grid": {
          "color": "hsl(--secondary)",
        },
        "max": "now",
        "min": 0,
        "ticks": {
          "autoSkipPadding": 20,
          "maxRotation": 0,
        },
        "time": {
          "displayFormats": {
            "day": "DD.MM",
            "hour": "HH:00",
            "minute": "HH:mm",
          },
          "minUnit": "minute",
          "tooltipFormat": "D.MM.YYYY",
        },
        "type": "time",
      },
      "y": {
        "grid": {
          "color": "hsl(--secondary)",
        },
        "min": 0,
      },
    },
  },
  "type": "line",
}
`;

exports[`dashboard / chart builders > for dark theme > should build activeUsers chart without data 1`] = `
{
  "data": {
    "datasets": [],
    "labels": [],
  },
  "options": {
    "plugins": {
      "annotation": {
        "annotations": {
          "annotation": {
            "borderColor": "hsl(--foreground)",
            "borderDash": [
              6,
              6,
            ],
            "borderDashOffset": 0,
            "borderWidth": 2,
            "label": {
              "content": [Function],
              "display": true,
              "position": "end",
            },
            "scaleID": "y",
            "type": "line",
            "value": [Function],
          },
        },
      },
    },
    "scales": {
      "x": {
        "grid": {
          "color": "hsl(--secondary)",
        },
        "max": "now",
        "min": 0,
        "ticks": {
          "autoSkipPadding": 20,
          "maxRotation": 0,
        },
        "time": {
          "displayFormats": {
            "day": "DD.MM",
            "hour": "HH:00",
            "minute": "HH:mm",
          },
          "minUnit": "minute",
          "tooltipFormat": "D.MM.YYYY",
        },
        "type": "time",
      },
      "y": {
        "grid": {
          "color": "hsl(--secondary)",
        },
        "min": 0,
      },
    },
  },
  "type": "line",
}
`;

exports[`dashboard / chart builders > for dark theme > should build time chart with data 1`] = `
{
  "data": {
    "datasets": [
      {
        "data": [
          3,
        ],
        "label": "chart 1",
        "pointHitRadius": 10,
      },
      {
        "data": [],
        "label": "chart 2",
        "pointHitRadius": 10,
      },
    ],
    "labels": [
      2,
    ],
  },
  "options": {
    "plugins": {
      "annotation": {
        "annotations": {
          "annotation": {
            "borderColor": "hsl(--foreground)",
            "borderDash": [
              6,
              6,
            ],
            "borderDashOffset": 0,
            "borderWidth": 2,
            "label": {
              "content": [Function],
              "display": true,
              "position": "end",
            },
            "scaleID": "y",
            "type": "line",
            "value": [Function],
          },
        },
      },
    },
    "scales": {
      "x": {
        "grid": {
          "color": "hsl(--secondary)",
        },
        "max": "now",
        "min": 0,
        "ticks": {
          "autoSkipPadding": 20,
          "maxRotation": 0,
        },
        "time": {
          "displayFormats": {
            "day": "DD.MM",
            "hour": "HH:00",
            "minute": "HH:mm",
          },
          "minUnit": "minute",
          "tooltipFormat": "D.MM.YYYY H:mm:ss.SSS",
        },
        "type": "time",
      },
      "y": {
        "grid": {
          "color": "hsl(--secondary)",
        },
        "min": 0,
        "title": {
          "display": true,
          "text": "мс",
        },
      },
    },
  },
  "type": "line",
}
`;

exports[`dashboard / chart builders > for dark theme > should build time chart without data 1`] = `
{
  "data": {
    "datasets": [],
    "labels": [],
  },
  "options": {
    "plugins": {
      "annotation": {
        "annotations": {
          "annotation": {
            "borderColor": "hsl(--foreground)",
            "borderDash": [
              6,
              6,
            ],
            "borderDashOffset": 0,
            "borderWidth": 2,
            "label": {
              "content": [Function],
              "display": true,
              "position": "end",
            },
            "scaleID": "y",
            "type": "line",
            "value": [Function],
          },
        },
      },
    },
    "scales": {
      "x": {
        "grid": {
          "color": "hsl(--secondary)",
        },
        "max": "now",
        "min": 0,
        "ticks": {
          "autoSkipPadding": 20,
          "maxRotation": 0,
        },
        "time": {
          "displayFormats": {
            "day": "DD.MM",
            "hour": "HH:00",
            "minute": "HH:mm",
          },
          "minUnit": "minute",
          "tooltipFormat": "D.MM.YYYY H:mm:ss.SSS",
        },
        "type": "time",
      },
      "y": {
        "grid": {
          "color": "hsl(--secondary)",
        },
        "min": 0,
        "title": {
          "display": true,
          "text": "мс",
        },
      },
    },
  },
  "type": "line",
}
`;

exports[`dashboard / chart builders > for light theme > should build activeUsers chart with data 1`] = `
{
  "data": {
    "datasets": [
      {
        "data": [
          3,
        ],
        "label": "chart 1",
        "pointHitRadius": 10,
      },
      {
        "data": [],
        "label": "chart 2",
        "pointHitRadius": 10,
      },
    ],
    "labels": [
      2,
    ],
  },
  "options": {
    "plugins": {
      "annotation": {
        "annotations": {
          "annotation": {
            "borderColor": "hsl(--foreground)",
            "borderDash": [
              6,
              6,
            ],
            "borderDashOffset": 0,
            "borderWidth": 2,
            "label": {
              "content": [Function],
              "display": true,
              "position": "end",
            },
            "scaleID": "y",
            "type": "line",
            "value": [Function],
          },
        },
      },
    },
    "scales": {
      "x": {
        "grid": {
          "color": "rgba(0,0,0,0.1)",
        },
        "max": "now",
        "min": 0,
        "ticks": {
          "autoSkipPadding": 20,
          "maxRotation": 0,
        },
        "time": {
          "displayFormats": {
            "day": "DD.MM",
            "hour": "HH:00",
            "minute": "HH:mm",
          },
          "minUnit": "minute",
          "tooltipFormat": "D.MM.YYYY",
        },
        "type": "time",
      },
      "y": {
        "grid": {
          "color": "rgba(0,0,0,0.1)",
        },
        "min": 0,
      },
    },
  },
  "type": "line",
}
`;

exports[`dashboard / chart builders > for light theme > should build activeUsers chart without data 1`] = `
{
  "data": {
    "datasets": [],
    "labels": [],
  },
  "options": {
    "plugins": {
      "annotation": {
        "annotations": {
          "annotation": {
            "borderColor": "hsl(--foreground)",
            "borderDash": [
              6,
              6,
            ],
            "borderDashOffset": 0,
            "borderWidth": 2,
            "label": {
              "content": [Function],
              "display": true,
              "position": "end",
            },
            "scaleID": "y",
            "type": "line",
            "value": [Function],
          },
        },
      },
    },
    "scales": {
      "x": {
        "grid": {
          "color": "rgba(0,0,0,0.1)",
        },
        "max": "now",
        "min": 0,
        "ticks": {
          "autoSkipPadding": 20,
          "maxRotation": 0,
        },
        "time": {
          "displayFormats": {
            "day": "DD.MM",
            "hour": "HH:00",
            "minute": "HH:mm",
          },
          "minUnit": "minute",
          "tooltipFormat": "D.MM.YYYY",
        },
        "type": "time",
      },
      "y": {
        "grid": {
          "color": "rgba(0,0,0,0.1)",
        },
        "min": 0,
      },
    },
  },
  "type": "line",
}
`;

exports[`dashboard / chart builders > for light theme > should build time chart with data 1`] = `
{
  "data": {
    "datasets": [
      {
        "data": [
          3,
        ],
        "label": "chart 1",
        "pointHitRadius": 10,
      },
      {
        "data": [],
        "label": "chart 2",
        "pointHitRadius": 10,
      },
    ],
    "labels": [
      2,
    ],
  },
  "options": {
    "plugins": {
      "annotation": {
        "annotations": {
          "annotation": {
            "borderColor": "hsl(--foreground)",
            "borderDash": [
              6,
              6,
            ],
            "borderDashOffset": 0,
            "borderWidth": 2,
            "label": {
              "content": [Function],
              "display": true,
              "position": "end",
            },
            "scaleID": "y",
            "type": "line",
            "value": [Function],
          },
        },
      },
    },
    "scales": {
      "x": {
        "grid": {
          "color": "rgba(0,0,0,0.1)",
        },
        "max": "now",
        "min": 0,
        "ticks": {
          "autoSkipPadding": 20,
          "maxRotation": 0,
        },
        "time": {
          "displayFormats": {
            "day": "DD.MM",
            "hour": "HH:00",
            "minute": "HH:mm",
          },
          "minUnit": "minute",
          "tooltipFormat": "D.MM.YYYY H:mm:ss.SSS",
        },
        "type": "time",
      },
      "y": {
        "grid": {
          "color": "rgba(0,0,0,0.1)",
        },
        "min": 0,
        "title": {
          "display": true,
          "text": "мс",
        },
      },
    },
  },
  "type": "line",
}
`;

exports[`dashboard / chart builders > for light theme > should build time chart without data 1`] = `
{
  "data": {
    "datasets": [],
    "labels": [],
  },
  "options": {
    "plugins": {
      "annotation": {
        "annotations": {
          "annotation": {
            "borderColor": "hsl(--foreground)",
            "borderDash": [
              6,
              6,
            ],
            "borderDashOffset": 0,
            "borderWidth": 2,
            "label": {
              "content": [Function],
              "display": true,
              "position": "end",
            },
            "scaleID": "y",
            "type": "line",
            "value": [Function],
          },
        },
      },
    },
    "scales": {
      "x": {
        "grid": {
          "color": "rgba(0,0,0,0.1)",
        },
        "max": "now",
        "min": 0,
        "ticks": {
          "autoSkipPadding": 20,
          "maxRotation": 0,
        },
        "time": {
          "displayFormats": {
            "day": "DD.MM",
            "hour": "HH:00",
            "minute": "HH:mm",
          },
          "minUnit": "minute",
          "tooltipFormat": "D.MM.YYYY H:mm:ss.SSS",
        },
        "type": "time",
      },
      "y": {
        "grid": {
          "color": "rgba(0,0,0,0.1)",
        },
        "min": 0,
        "title": {
          "display": true,
          "text": "мс",
        },
      },
    },
  },
  "type": "line",
}
`;

exports[`dashboard / chart builders > should build authenticationFunnel chart with data 1`] = `
{
  "data": {
    "datasets": [
      {
        "backgroundColor": [
          "--success-color",
          "--error-color",
        ],
        "data": [
          0.5,
          0.5,
        ],
      },
    ],
    "labels": [
      "аутентификации",
      "уходы",
    ],
  },
  "options": {
    "borderColor": "hsl(--background)",
    "cutout": "50%",
    "plugins": {
      "legend": {
        "display": false,
      },
      "tooltip": {
        "callbacks": {
          "label": [Function],
        },
      },
    },
  },
  "type": "doughnut",
}
`;

exports[`dashboard / chart builders > should build authenticationFunnel chart without data 1`] = `
{
  "data": {
    "datasets": [
      {
        "backgroundColor": [
          "--success-color",
          "--error-color",
        ],
        "data": [
          1,
        ],
      },
    ],
    "labels": [
      "аутентификации",
      "уходы",
    ],
  },
  "options": {
    "borderColor": "--success-color",
    "cutout": "100%",
    "plugins": {
      "legend": {
        "display": false,
      },
      "tooltip": {
        "callbacks": {
          "label": [Function],
        },
      },
    },
  },
  "type": "doughnut",
}
`;

exports[`dashboard / chart builders > should build successRate chart with data 1`] = `
{
  "data": {
    "datasets": [
      {
        "backgroundColor": [
          "--success-color",
          "--error-color",
        ],
        "data": [
          0.5,
          0.5,
        ],
      },
    ],
    "labels": [
      "success rate",
      "error rate",
    ],
  },
  "options": {
    "borderColor": "hsl(--background)",
    "cutout": "50%",
    "plugins": {
      "legend": {
        "display": false,
      },
      "tooltip": {
        "callbacks": {
          "label": [Function],
        },
      },
    },
  },
  "type": "doughnut",
}
`;

exports[`dashboard / chart builders > should build successRate chart without data 1`] = `
{
  "data": {
    "datasets": [
      {
        "backgroundColor": [
          "--success-color",
          "--error-color",
        ],
        "data": [
          1,
        ],
      },
    ],
    "labels": [
      "success rate",
      "error rate",
    ],
  },
  "options": {
    "borderColor": "--success-color",
    "cutout": "100%",
    "plugins": {
      "legend": {
        "display": false,
      },
      "tooltip": {
        "callbacks": {
          "label": [Function],
        },
      },
    },
  },
  "type": "doughnut",
}
`;
